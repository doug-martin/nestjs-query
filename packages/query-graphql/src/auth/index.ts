export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
