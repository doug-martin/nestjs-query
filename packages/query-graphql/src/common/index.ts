export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
