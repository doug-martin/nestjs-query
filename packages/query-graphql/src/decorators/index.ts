export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
