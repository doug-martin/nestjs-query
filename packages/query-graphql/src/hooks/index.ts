export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
