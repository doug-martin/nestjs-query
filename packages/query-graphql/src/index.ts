export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
