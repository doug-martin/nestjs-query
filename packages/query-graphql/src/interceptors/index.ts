export * from './hook.interceptor';
export * from './authorizer.interceptor';
