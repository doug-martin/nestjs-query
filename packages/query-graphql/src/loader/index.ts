export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
