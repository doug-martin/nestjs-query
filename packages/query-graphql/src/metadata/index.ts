import { GraphQLQueryMetadataStorage } from './metadata-storage';

export { FilterableFieldDescriptor } from './metadata-storage';

/** @internal */
let storage: GraphQLQueryMetadataStorage;
/** @internal */
export function getMetadataStorage(): GraphQLQueryMetadataStorage {
  if (!storage) {
    storage = new GraphQLQueryMetadataStorage();
  }
  return storage;
}
