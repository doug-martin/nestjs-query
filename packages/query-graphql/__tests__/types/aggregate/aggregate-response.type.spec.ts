// eslint-disable-next-line max-classes-per-file
import { AggregateResponse } from '@nestjs-query/core';
import { Resolver, Query, ObjectType, GraphQLISODateTime } from '@nestjs/graphql';
import { AggregateResponseType, FilterableField } from '../../../src';
import { generateSchema } from '../../__fixtures__';

describe('AggregateResponseType', (): void => {
  @ObjectType()
  class FakeType {
    @FilterableField()
    stringField!: string;

    @FilterableField()
    numberField!: number;

    @FilterableField()
    boolField!: boolean;

    @FilterableField(() => GraphQLISODateTime)
    dateField!: Date;
  }

  it('should create an aggregate type with the correct fields for each type', async () => {
    const AggResponse = AggregateResponseType(FakeType);
    @Resolver()
    class AggregateResponseTypeSpec {
      @Query(() => AggResponse)
      aggregate(): AggregateResponse<FakeType> {
        return {};
      }
    }
    const schema = await generateSchema([AggregateResponseTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should return the same class if called multiple times', async () => {
    AggregateResponseType(FakeType);
    const AggResponse = AggregateResponseType(FakeType);
    @Resolver()
    class AggregateResponseTypeSpec {
      @Query(() => AggResponse)
      aggregate(): AggregateResponse<FakeType> {
        return {};
      }
    }
    const schema = await generateSchema([AggregateResponseTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an aggregate type with a custom name', async () => {
    const AggResponse = AggregateResponseType(FakeType, { prefix: 'CustomPrefix' });
    @Resolver(() => AggResponse)
    class AggregateResponseTypeSpec {
      @Query(() => AggResponse)
      aggregate(): AggregateResponse<FakeType> {
        return {};
      }
    }
    const schema = await generateSchema([AggregateResponseTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('throw an error if the type is not registered', () => {
    class BadType {
      id!: number;
    }
    expect(() => AggregateResponseType(BadType)).toThrow(
      'Unable to make AggregationResponseType. Ensure BadType is annotated with @nestjs/graphql @ObjectType',
    );
  });
  it('throw an error if fields are not found', () => {
    @ObjectType()
    class BadType {
      id!: number;
    }
    expect(() => AggregateResponseType(BadType)).toThrow(
      'No fields found to create AggregationResponseType for BadType. Ensure fields are annotated with @FilterableField',
    );
  });
});
