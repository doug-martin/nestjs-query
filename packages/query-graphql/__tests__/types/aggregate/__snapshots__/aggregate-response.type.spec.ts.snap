// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateResponseType should create an aggregate type with a custom name 1`] = `
type FakeTypeAggregateGroupBy {
  stringField: String
  numberField: Float
  boolField: Boolean
  dateField: DateTime
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type FakeTypeCountAggregate {
  stringField: Int
  numberField: Int
  boolField: Int
  dateField: Int
}

type FakeTypeSumAggregate {
  numberField: Float
}

type FakeTypeAvgAggregate {
  numberField: Float
}

type FakeTypeMinAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type FakeTypeMaxAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type CustomPrefixAggregateGroupBy {
  stringField: String
  numberField: Float
  boolField: Boolean
  dateField: DateTime
}

type CustomPrefixCountAggregate {
  stringField: Int
  numberField: Int
  boolField: Int
  dateField: Int
}

type CustomPrefixSumAggregate {
  numberField: Float
}

type CustomPrefixAvgAggregate {
  numberField: Float
}

type CustomPrefixMinAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type CustomPrefixMaxAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type CustomPrefixAggregateResponse {
  groupBy: CustomPrefixAggregateGroupBy
  count: CustomPrefixCountAggregate
  sum: CustomPrefixSumAggregate
  avg: CustomPrefixAvgAggregate
  min: CustomPrefixMinAggregate
  max: CustomPrefixMaxAggregate
}

type Query {
  aggregate: CustomPrefixAggregateResponse!
}

`;

exports[`AggregateResponseType should create an aggregate type with the correct fields for each type 1`] = `
type FakeTypeAggregateGroupBy {
  stringField: String
  numberField: Float
  boolField: Boolean
  dateField: DateTime
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type FakeTypeCountAggregate {
  stringField: Int
  numberField: Int
  boolField: Int
  dateField: Int
}

type FakeTypeSumAggregate {
  numberField: Float
}

type FakeTypeAvgAggregate {
  numberField: Float
}

type FakeTypeMinAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type FakeTypeMaxAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type FakeTypeAggregateResponse {
  groupBy: FakeTypeAggregateGroupBy
  count: FakeTypeCountAggregate
  sum: FakeTypeSumAggregate
  avg: FakeTypeAvgAggregate
  min: FakeTypeMinAggregate
  max: FakeTypeMaxAggregate
}

type Query {
  aggregate: FakeTypeAggregateResponse!
}

`;

exports[`AggregateResponseType should return the same class if called multiple times 1`] = `
type FakeTypeAggregateGroupBy {
  stringField: String
  numberField: Float
  boolField: Boolean
  dateField: DateTime
}

"""
A date-time string at UTC, such as 2019-12-03T09:54:33Z, compliant with the date-time format.
"""
scalar DateTime

type FakeTypeCountAggregate {
  stringField: Int
  numberField: Int
  boolField: Int
  dateField: Int
}

type FakeTypeSumAggregate {
  numberField: Float
}

type FakeTypeAvgAggregate {
  numberField: Float
}

type FakeTypeMinAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type FakeTypeMaxAggregate {
  stringField: String
  numberField: Float
  dateField: DateTime
}

type FakeTypeAggregateResponse {
  groupBy: FakeTypeAggregateGroupBy
  count: FakeTypeCountAggregate
  sum: FakeTypeSumAggregate
  avg: FakeTypeAvgAggregate
  min: FakeTypeMinAggregate
  max: FakeTypeMaxAggregate
}

type Query {
  aggregate: FakeTypeAggregateResponse!
}

`;
